src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/action/SnapshotAction.kt

```diff
@@ -0,0 +1,36 @@
+package com.amazon.opendistroforelasticsearch.indexstatemanagement.action
+
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.ManagedIndexMetaData
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.action.ActionConfig.ActionType
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.action.SnapshotActionConfig
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.Step
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.snapshot.AttemptSnapshotStep
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.snapshot.WaitForSnapshotStep
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+
+class SnapshotAction(
+    clusterService: ClusterService,
+    client: Client,
+    managedIndexMetaData: ManagedIndexMetaData,
+    config: SnapshotActionConfig
+) : Action(ActionType.SNAPSHOT, config, managedIndexMetaData) {
+    private val attemptSnapshotStep = AttemptSnapshotStep(clusterService, client, config, managedIndexMetaData)
+    private val waitForSnapshotStep = WaitForSnapshotStep(clusterService, client, config, managedIndexMetaData)
+
+    private val stepNameToStep: LinkedHashMap<String, Step> = linkedMapOf(
+        AttemptSnapshotStep.name to attemptSnapshotStep,
+        WaitForSnapshotStep.name to waitForSnapshotStep
+    )
+
+    override fun getSteps(): List<Step> = stepNameToStep.values.toList()
+
+    @Suppress("ReturnCount")
+    override fun getStepToExecute(): Step {
```

Comment:
Seems like this doesn't handle the case where the step is still on waitForSnapshotStep? Looks like it might return attemptSnapshotStep instead.
i.e. the snapshot is taking a while to complete for a large index and it will take multiple executions for waitFor to successfully complete.

Also if the step was on AttempSnapshotStep but failed/was retried. Seems like it will move forward to waitForSnapshotStep?
I'm assuming we only want to move to waitFor step current step is AttemptSnapshotStep and stepMetaData.stepStatus has a status of COMPLETED.

Reply:
Updated, I've changed the logic to suit these cases.